### DISCUSSION TITLE
Testing GitHub Actions?

### DISCUSSION THREAD

**Discussion by tomschr** [community/community#25754] [Actions]
<p>In <a href="https://github.com/github/docs/discussions/2501" rel="noopener nofollow ugc">discussions#2501</a> I’ve asked to amend the GitHub documentation with a new topic about testing GitHub Actions (GHA). Martin Lopes suggested to ask here. :slight_smile:</p>
<p>However, before we can document something, we need to clarify <strong>how</strong> can we test GHA. I’m not asking to debug them, this is only partly related. The question here is how do we evaluate GHA <em>locally</em> before it’s evaluated on GitHub itself.</p>
<p>It is really cumbersome to open pull requests etc. just to see if the event is evoked. It takes time, is error-prone, and a cumbersome step-by-step process.</p>
<p>Any suggestions/ideas/recommendations?</p>
<p>Thanks! :+1:</p>

**Comment by lucernae** [comment]
<p>I literally create a sandbox repo where I can execute action and see the workflow context myself.</p>
<p>It will be nice if there is a better way. I think the reason why we can’t easily test action yet is because the api event invoke comes from Github. If we want to test it ourselves we need to emulate the event somehow and it will get messy.</p>

**Comment by jsoref** [comment, marked as answer]
<p>Depending on how advanced your workflow is, you might be able to use:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/nektos/act">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/nektos/act" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/2119212/181093059-885ff3d9-707c-4299-b247-e4264a025b78.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/5/c58ae5451c1f4a51dc43b5adb188a9b3a19bb877_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/c/5/c58ae5451c1f4a51dc43b5adb188a9b3a19bb877_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/c/5/c58ae5451c1f4a51dc43b5adb188a9b3a19bb877.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/5/c58ae5451c1f4a51dc43b5adb188a9b3a19bb877_2_10x10.png"></div>

<h3><a href="https://github.com/nektos/act" target="_blank" rel="noopener nofollow ugc">GitHub - nektos/act: Run your GitHub Actions locally 🚀</a></h3>

  <p>Run your GitHub Actions locally 🚀. Contribute to nektos/act development by creating an account on GitHub.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
